code/modules/admin/verbs/map_export.dm

```diff
@@ -234,65 +259,50 @@ GLOBAL_LIST_INIT(save_file_chars, list(
 						empty = FALSE
 				current_header += "[empty ? "" : ",\n"][place],\n[location])\n"
 				//====Fill the contents file====
-				//Compression is done here
-				var/position_of_header = header_dat.Find(current_header)
-				if(position_of_header)
-					//If the header has already been saved, change the character to the other saved header
-					header_char = header_chars[position_of_header]
-				else
-					header += "\"[header_char]\" = [current_header]"
-					header_chars += header_char
-					header_dat += current_header
-					index ++
-				contents += "[header_char]\n"
+				var/textiftied_header = current_header.Join()
+				// If we already know this header just use its key, otherwise we gotta make a new one
+				var/key = header_data[textiftied_header]
+				if(!key)
+					key = calculate_tgm_header_index(key_index, layers)
+					key_index++
+					header += "\"[key]\" = [textiftied_header]"
+					header_data[textiftied_header] = key
+				contents += "[key]\n"
 			contents += "\"}"
-	return "//[DMM2TGM_MESSAGE]\n[header][contents]"
+	return "//[DMM2TGM_MESSAGE]\n[header.Join()][contents.Join()]"
 
-//vars_to_save = list() to save all vars
 /proc/generate_tgm_metadata(atom/object)
-	var/dat = ""
-	var/data_to_add = list()
+	var/list/data_to_add = list()
+
 	var/list/vars_to_save = object.get_save_vars()
-	if(!vars_to_save)
-		return
-	for(var/variable in object.vars)
+	for(var/variable in vars_to_save)
 		CHECK_TICK
-		if(!(variable in vars_to_save))
-			continue
 		var/value = object.vars[variable]
 		if(!value)
```

Comment:
I see a small problem here. Aargs with a value of 0, "" or null won't be saved.

Reply:
oh right, yeah.
IDK why it's written like this, it's dumb
We really want (is the variable in vars) but that's slow as hell, maybe just write anyways idk

Reply:
Yeah, just write it. Negation values are valid values by `tgm_encode` standards, while real bad ones are things like live datums.

Reply:
I will use BOTH. I do not know if the map reader supports non existing vars and I don't really want to encode them.
this should be cheap enough to be fine

Reply:
nevermind that is WAY too expensive. we'll just treat em all as valid and cry